test/state/system_contracts.cpp

```diff
@@ -51,25 +67,52 @@ StateDiff system_call_block_start(const StateView& state_view, const BlockInfo&
         // > if no code exists at [address], the call must fail silently.
         const auto code = state_view.get_account_code(addr);
         if (code.empty())
+        {
+            requests.emplace_back(request_type);
             continue;
+        }
 
-        const auto input = get_input(block, block_hashes);
+        bytes32 input32;
+        bytes_view input;
+        if (get_input != nullptr)
+        {
+            input32 = get_input(block, block_hashes);
+            input = input32;
+        }
 
         const evmc_message msg{
             .kind = EVMC_CALL,
             .gas = 30'000'000,
             .recipient = addr,
             .sender = SYSTEM_ADDRESS,
-            .input_data = input.bytes,
-            .input_size = std::size(input.bytes),
+            .input_data = input.data(),
+            .input_size = input.size(),
         };
 
         const Transaction empty_tx{};
         Host host{rev, vm, state, block, block_hashes, empty_tx};
         [[maybe_unused]] const auto res = vm.execute(host, rev, msg, code.data(), code.size());
         assert(res.status_code == EVMC_SUCCESS);
+        requests.emplace_back(request_type, bytes(res.output_data, res.output_size));
     }
+
     // TODO: Should we return empty diff if no system contracts?
-    return state.build_diff(rev);
+    return {state.build_diff(rev), requests};
+}
+}  // namespace
+
+StateDiff system_call_block_start(const StateView& state_view, const BlockInfo& block,
+    const BlockHashes& block_hashes, evmc_revision rev, evmc::VM& vm)
+{
+    // No requests are generated in block start system calls.
+    const auto [state_diff, _] =
+        system_call(SYSTEM_CONTRACTS_BLOCK_START, state_view, block, block_hashes, rev, vm);
+    return state_diff;
+}
+
+std::pair<StateDiff, std::vector<Requests>> system_call_block_end(const StateView& state_view,
```

Comment:
Does the EIP-7002 make changes to the state?

Reply:
Yes it does, and 7251 too.

For both there is some kind of queue of requests stored in a contract, and system call dequeues items.